packages/zoltan2/test/color/ReadMatrixFromFile.hpp

```diff
@@ -0,0 +1,325 @@
+	  local_ordinal_type curPos = rowPtr[theirRow];
+
+	  if (curNumEntries > 0) {
+
+	    for(local_ordinal_type ii = 0; ii < curNumEntries; ++ii) {
+	      theirColInd[theirCurPos++] = colInd[curPos++];
+	    }
+	    
+	    //create theirValues if you are going to send
+
+	  }
+	}
+
+	Teuchos::send (*pComm, static_cast<int> (theirNumEntries), theirColInd.getRawPtr(), p);
+	// send (*pComm, static_cast<int> (theirNumEntries),
+	//       theirValues.getRawPtr(), p);
+
+	if (debug) {
+	  std::cout << "-- Proc 0: Finished with proc " << p << std::endl;
+	}
+	
+      } // If proc p owns at least one row
+
+      
+    } // For each proc p not the root proc 0
+  } // If I'm (not) the root proc 0
+
+  if (debug && myRank == 0) {
+    std::cout << "-- Proc 0: About to fill in myRowPtr" << std::endl;
+  }
+
+  myRowPtr = Teuchos::ArrayRCP<size_t> (myNumRows+1);
+  myRowPtr[0] = 0;
+  for (size_t k = 1; k < myNumRows+1; ++k) {
+    myRowPtr[k] = myRowPtr[k-1] + myNumEntriesPerRow[k-1];
+  }
+  if (debug && myRank == 0) {
+    std::cout << "-- Proc 0: Done with distribute" << std::endl;
+  }
+
+
+}
+
+template <typename crs_matrix_type>
+Teuchos::RCP<crs_matrix_type>
+readBinaryFile(std::string filename, const Teuchos::RCP<const Teuchos::Comm<int>> pComm, bool callFillComplete=true, bool debug=false)
+{
+
+  typedef typename crs_matrix_type::global_ordinal_type global_ordinal_type;
+  typedef typename crs_matrix_type::local_ordinal_type local_ordinal_type;
+  typedef typename crs_matrix_type::scalar_type scalar_type;
+  typedef typename crs_matrix_type::node_type node_type;
+
+  typedef typename Tpetra::Map<local_ordinal_type, global_ordinal_type, node_type> map_type;
+
+  const int myRank = pComm->getRank();
+  const int rootRank = 0;
+
+  if (debug && myRank == rootRank) {
+    std::cout << "Binary CRS reader: readSparse:" << std::endl
+  	      << "-- Reading started" << std::endl;
+  }
+
+  Teuchos::RCP<std::ifstream> in;
+  global_ordinal_type globalNumRows;
+  global_ordinal_type globalNumNonzeros;
+  
+  if (myRank == rootRank) {
+    
+    // Open the file
+    in = Teuchos::RCP<std::ifstream>(new std::ifstream(filename, std::ios::in | std::ios::binary));
+
+    // Read number of vertices and number of edges
+    in->read((char *)&globalNumRows, sizeof(global_ordinal_type));
+    in->read((char *)&globalNumNonzeros, sizeof(global_ordinal_type));
+
+    TEUCHOS_TEST_FOR_EXCEPTION(globalNumRows <= 0 || globalNumNonzeros <= 0, std::invalid_argument,
+  			       "Global number of rows or nonzeros have nonpositive value." << globalNumRows << " " << globalNumNonzeros << " " << sizeof(global_ordinal_type) );
+
+
+  }
+  
+  broadcast (*pComm, rootRank, 1, &globalNumRows);
+  broadcast (*pComm, rootRank, 1, &globalNumNonzeros);
+
+  //std::cout << globalNumRows << " " << globalNumNonzeros << std::endl;
+
+  global_ordinal_type *rowPtr;
+  global_ordinal_type *colInd;
+
+  if (myRank == rootRank) {
+
+    rowPtr = new global_ordinal_type[globalNumRows+1];
```

Comment:
Never use `new T[N]` without at least wrapping it in `unique_ptr` or `shared_ptr`.  You can use `unique_ptr` here:
```cpp
std::unique_ptr<global_ordinal_type> rowPtr;
if (myRank == rootRank) {
  rowPtr = std::unique_ptr<global_ordinal_type>(new global_ordinal_type[globalNumNonzeros]);
}
```

Reply:
Done.